pkg/streamingpromql/operators/aggregation.go

```diff
@@ -152,61 +156,99 @@ func (a *Aggregation) NextSeries(ctx context.Context) (types.InstantVectorSeries
 		thisSeriesGroup := a.remainingInnerSeriesToGroup[0]
 		a.remainingInnerSeriesToGroup = a.remainingInnerSeriesToGroup[1:]
 
-		if thisSeriesGroup.sums == nil {
+		if len(s.Floats) > 0 && thisSeriesGroup.floatSums == nil {
 			// First series for this group, populate it.
 
-			thisSeriesGroup.sums, err = a.Pool.GetFloatSlice(steps)
+			thisSeriesGroup.floatSums, err = a.Pool.GetFloatSlice(steps)
 			if err != nil {
 				return types.InstantVectorSeriesData{}, err
 			}
 
-			thisSeriesGroup.present, err = a.Pool.GetBoolSlice(steps)
+			thisSeriesGroup.floatPresent, err = a.Pool.GetBoolSlice(steps)
 			if err != nil {
 				return types.InstantVectorSeriesData{}, err
 			}
+			thisSeriesGroup.floatSums = thisSeriesGroup.floatSums[:steps]
+			thisSeriesGroup.floatPresent = thisSeriesGroup.floatPresent[:steps]
+		}
 
-			thisSeriesGroup.sums = thisSeriesGroup.sums[:steps]
-			thisSeriesGroup.present = thisSeriesGroup.present[:steps]
+		if len(s.Histograms) > 0 && thisSeriesGroup.histogramValues == nil {
+			// First series for this group, populate it.
+			thisSeriesGroup.histogramValues, err = a.Pool.GetHistogramPointerSlice(steps)
+			if err != nil {
+				return types.InstantVectorSeriesData{}, err
+			}
+			thisSeriesGroup.histogramValues = thisSeriesGroup.histogramValues[:steps]
 		}
 
 		for _, p := range s.Floats {
 			idx := (p.T - start) / interval
-			thisSeriesGroup.sums[idx] += p.F
-			thisSeriesGroup.present[idx] = true
+			thisSeriesGroup.floatSums[idx] += p.F
+			if !thisSeriesGroup.floatPresent[idx] {
+				thisSeriesGroup.floatPointCount += 1
+			}
+			thisSeriesGroup.floatPresent[idx] = true
+		}
+
+		for _, p := range s.Histograms {
+			idx := (p.T - start) / interval
+			if thisSeriesGroup.histogramValues[idx] == nil {
+				// We copy here because we modify the histogram through Add later on.
+				thisSeriesGroup.histogramValues[idx] = p.H.Copy()
```

Comment:
I don't believe this needs to be copied - it doesn't matter if we modify the histogram, nothing else should use it.

Reply:
The tests fail without copying this.

I need to verify what else is using it.

However, as we support more complex functions/operations, would it not be possible that they would also need access to the original series?

Reply:
>However, as we support more complex functions/operations, would it not be possible that they would also need access to the original series?

By default, nothing should be retaining the original series - the caller of `NextSeries()` should be the only one with access to the series returned by `NextSeries()`, and nothing else should modify that series.

If the caller of `NextSeries()` needs to retain the original value as well as a modified copy, then it should be up to that caller to copy it - but under normal circumstances I wouldn't want this to be a thing, as all the copies will mean more allocations and therefore impact performance.

Reply:
So without the Copy the tests fail when doing a range query (which is done as a comparison to the instant query https://github.com/grafana/mimir/blob/main/vendor/github.com/prometheus/prometheus/promql/promqltest/test.go#L881).

This is because the range query will lookback to get a point, and then add that point onto the current position. However, the lookback gets the modified Histogram so the results end up being wrong.

If we have a way of telling there will be no lookback then I think we can get away with no copy. Otherwise, I do think we need to do a copy here which will unfortunately have an impact of performance and memory.

I suspect this will be very difficult to achieve without complex changes to the memoized code.

Reply:
Resolving as conversation continued below https://github.com/grafana/mimir/pull/8360#discussion_r1648309497